### ISSUE TITLE
Use Case: DSTL01-TRANSFORMATION-MATRIX

### ISSUE THREAD

**Issue by paulc-dstl** [opengeospatial/ogc-geosparql#604]
# Use Case DSTL01

## Transformation Matrices

As a GeoSPARQL data user I would like to transform geometry coordinates using a matrix. Matrix operations can be defined using 2x2 or 3x3 matrices for two- and three-dimensional transformations respectively. Providing a generic matrix based transformation function allows for a comprehensive range of basic geometry operations in 2D or 3D. A generic matrix transformation can be parameterized and used as the basis for a wide range of other functions. Matrices can be multiplied to provide compound transformations. 

Transformations can be considered simple and working in cartesian 2D/3D coordinate space only. For more complex transformations on geometries, use eg [PROJ](https://proj.org/en/stable/operations/transformations/index.html). 

A matrix based transformation can be used to:

- Translate in two or three dimensions
- Scale along one, two or three axes
- Rotate about any point or axis
- Shear along any axis
- Reflect in any line (2d) or plane (3d)
- Project
- Compound transformation comprising any of the above
### Examples:

- scale (in 2d) by k in the x-axis: 

$$\begin{bmatrix}k&0\\\0&1\end{bmatrix}$$

- scale (in 2d) by k in the y-axis: 

$$\begin{bmatrix}1&0\\\0&k\end{bmatrix}$$

- rotate by angle $$\Theta$$ about the origin:

$$\begin{bmatrix}\cos \theta& -\sin \theta\\\ \sin \theta & \cos \theta \end{bmatrix}$$

- rotate by angle $$\Theta$$ about an axis defined by a unit vector (x,y,z):

$$\begin{bmatrix}xx(1-\cos \theta)+\cos\theta & yx(1-\cos\theta)-z\sin\theta & zx(1-\cos\theta)+y\sin\theta\\\ xy(1-\cos\theta)+z\sin\theta & yy(1-\cos\theta)+\cos\theta & zy(1-\cos\theta)-x\sin\theta \\\ xz(1-\cos\theta)-y\sin\theta & yz(1-\cos\theta)+x\sin\theta & zz(1-\cos\theta)+\cos\theta \end{bmatrix}$$

## Function signature: 

`matrixTransform(geometry_in: ogc:geomLiteral, transform: ex:matrix): geometry_out: ogc:geomLiteral`

## Actor

GeoSPARQL data user (human or software agent querying RDF dataset with GeoSPARQL)

## Preconditions 

Geometries with 2D or 3D representations are present in a dataset to be queried. 
There exists some usable definition of a matrix datatype to be specified as the type to be used for the function parameter `transform`. This should be equivalent to the [numpy.array](https://numpy.org/doc/stable/reference/generated/numpy.array.html) or [JSON Schema](https://json-schema.org/understanding-json-schema/reference/array) datatypes: eg `[[1,0],[0,1]]` and should allow for the specification of MxN matrices, typically 2x2 or 3x3 matrices. 

## Postconditions

A SPARQL query can be executed that includes a function to return geometry with coordinates transformed from input geometry coordinate using a matrix which specifies a transformation.

## Steps

Actor: discovers the matrix transformation function
Actor: executes a SPARQL query containing a matrix transformation function with arguments for a) a geometry and b) a matrix representing the desired transformation.
System: returns a SPARQL result with a geometry object translated from the input geometry using the input matrix.

**Comment by paulc-dstl** [comment]
Looking at existing [use cases](https://github.com/opengeospatial/ogc-geosparql/labels/Use%20Case), this generic transformation matrix function can also support the following the following use cases:

- #602 
- #576 
- #559 
- #560 
- #547

**Comment by ar-chad** [comment]
Yes, so there will be question if, at least some of, those simpler functions should be still in GeoSPARQL or matrix transform would be the one to keep.

**Comment by paulc-dstl** [comment]
I think likely keep the simpler ones also - if a users wants to eg scale along z-axis, i imagine many users would prefer to use a simple dedicated function rather than work out a matrix transformation. Even if when it comes to implementation, the simpler functions become wrappers which call the matrix function.

**Comment by situx** [comment]
@paulc-dstl maybe I have missed it, but I think we would need to define a literal type for the matrix representation, correct?

**Comment by ar-chad** [comment]
In CityRDF there are two classes core:TransformationMatrix2x2 core:TransformationMatrix3x4. They are subclasses of core:DoubleList 

`core:DoubleList a owl:Class ;
    rdfs:label "DoubleList"@en ;
    rdfs:subClassOf [ a owl:Restriction ;
            owl:allValuesFrom xsd:double ;
            owl:onProperty core:list ],
        [ a owl:Restriction ;
            owl:onDataRange xsd:double ;
            owl:onProperty core:list ;
            owl:qualifiedCardinality "1"^^xsd:nonNegativeInteger ] ;
    skos:definition "DoubleList is an ordered sequence of double values."@en .`

@paulc-dstl do you think that this could be reused in any way for GeoSPARQL?

https://github.com/ogcincubator/cityrdf/blob/master/citygml-owl/CityRDF/core.ttl

**Comment by situx** [comment]
I tried to apply the concepts that @ar-chad pointed out in rdflib-geosparql, but figured that we actually need a datatype rather than an instance, so that was not an option.
For now I defined a geo:matrixLiteral with an example as follows:

```ttl
geofext:affineTransformation(?geom,"[1.5 1.0 0.0 1.0 2.0 -1.0]"^^geo:matrixLiteral)
```
The function would take either a matrix for 2D or 3D transformation 
- [a, b, d, e, xoff, yoff]
- [a, b, c, d, e, f, g, h, i, xoff, yoff, zoff]

I suggest we use that and we could be compatible with other transformation matrices that arise as well.
